### PR TITLE
Add: Static code upload challenge Model changes and Message Queue Changes 

### PR DESCRIPTION
**Pull request by savish28**
Context: Static Code upload challenge Addition
Overview: Model changes + Message Queue Changes
Deliverables:
- [x]  Add is_static_dataset_docker_based_challenge field in challenge model.
- [x]  Add submission_input_file field in submission model.
- [x] Add is_static_code_upload_submission flag in SQS queue message for submissions made to static code upload challenge.
- [x] Worker Changes w.r.t Message body.

@KhalidRmb 
cc: @Ram81 
This PR had changes in various files so, for abstraction would create a new PR for Django admin actions.

### CODE REVIEW

File: apps/jobs/models.py [hunk 1]
```diff
@@ -85,6 +91,12 @@ class Submission(TimeStampedModel):
     input_file = models.FileField(
         upload_to=RandomFileName("submission_files/submission_{id}")
     )
+    # Model to store submitted docker image information
+    docker_input_file = models.FileField(
```
Comment (Ram81) [inline apps/jobs/models.py:95]: @savish28 this should be `submission_input_file`. We are going to store the output of docker evaluation in a separate file and not maintain multiple copies of input file. Having different submission flows isn't ideal.

File: apps/jobs/utils.py [hunk 2]
```diff
@@ -211,19 +211,25 @@ def handle_submission_rerun(submission, updated_status):
         "challenge_pk": submission.challenge_phase.challenge.pk,
         "phase_pk": submission.challenge_phase.pk,
         "submission_pk": submission.pk,
+        "is_static_code_upload_submission": False,
     }
 
     if submission.challenge_phase.challenge.is_docker_based:
         try:
-            response = requests.get(submission.input_file.url)
+            response = requests.get(submission.docker_input_file.url)
```
Comment (Ram81) [inline apps/jobs/utils.py:219]: Please revert these changes too

File: scripts/workers/code_upload_submission_worker.py [hunk 3]
```diff
@@ -347,6 +347,11 @@ def main():
         message = evalai.get_message_from_sqs_queue()
         message_body = message.get("body")
         if message_body:
+            if (
+                challenge.is_static_dataset_docker_based_challenge
+                and not message_body.get("is_static_code_upload_submission")
+            ):
+                continue
```
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:354]: @savish28 let's not add these changes yet. We can these changes once we start working on the worker changes.

File: apps/jobs/views.py [hunk 4]
```diff
@@ -483,6 +489,18 @@ def change_submission_data_and_visibility(
     if serializer.is_valid():
         serializer.save()
         response_data = serializer.data
+        if (
+            request.FILES.get("input_file")
```
Comment (Ram81) [inline apps/jobs/views.py:493]: @savish28 this will be `submission_input_file`

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Comment by savish28** [comment]
@Ram81 used `submission_input_file`.

File: apps/jobs/views.py [hunk 5]
```diff
@@ -2308,10 +2323,12 @@ def get_github_badge_data(
...
             participant_team_pk
         ):
-            data["message"] = f"{challenge_obj.title} Rank #{idx+1}"
+            data["message"] = "{} Rank #{}".format(
+                challenge_obj.title, idx + 1
+            )
             break
         else:
-            data["message"] = f"{challenge_obj.title}"
+            data["message"] = "{}".format(challenge_obj.title)
```
Comment (KhalidRmb) [inline apps/jobs/views.py:2331]: Maybe just leave the title as is? 
`data["message"] = "{}".format(challenge_obj.title)` -> `data["message"] = challenge_obj.title`

Comment (savish28) [reply apps/jobs/views.py:2331]: @KhalidRmb This is a pre-commit change and unrelated with the pr. It was showing syntax error so I shifted to the format type. I think both the lines means same right?

File: apps/jobs/models.py [hunk 6]
```diff
@@ -85,6 +85,11 @@ class Submission(TimeStampedModel):
     input_file = models.FileField(
         upload_to=RandomFileName("submission_files/submission_{id}")
     )
+    submission_input_file = models.FileField(
+        upload_to=RandomFileName("submission_files/submission_{id}"),
```
Comment (KhalidRmb) [inline apps/jobs/models.py:89]: Is this right? @savish28  Why is the upload_to location same as that of input_file? Will only one of them be uploaded in any scenario?
cc: @Ram81

File: tests/unit/jobs/test_views.py [hunk 7]
```diff
@@ -669,6 +669,7 @@ def test_get_challenge_submissions(self):
                 "status": self.submission.status,
                 "input_file": "http://testserver%s"
                 % (self.submission.input_file.url),
+                "submission_input_file": None,
```
Comment (KhalidRmb) [inline tests/unit/jobs/test_views.py:672]: Can you follow similar format for checking this? i.e `self.submission.submission_input_file` ? @savish28 
Same in other places.
cc: @Ram81

File: apps/jobs/models.py (hunk 6 above)
Comment (savish28) [reply apps/jobs/models.py:89]: Actually, input_file would contain the docker image URI information in code-based challenges and submission_input_file would contain the output of the job created the code. All the submission-related file data is stored in submission_{id}, like stderr/stdout/input_file so I think this would be the best location to store the file.

File: apps/jobs/views.py (hunk 5 above)
Comment (KhalidRmb) [reply apps/jobs/views.py:2331]: I understand and they do mean the same, but using string formatting here is redundant. Better to remove it

File: apps/jobs/models.py (hunk 6 above)
Comment (KhalidRmb) [reply apps/jobs/models.py:89]: Got it. Thanks

File: tests/unit/jobs/test_views.py (hunk 7 above)
Comment (savish28) [reply tests/unit/jobs/test_views.py:672]: @KhalidRmb, Ideally since it is a file it should be "http://testserver%s"% (self.submission.submission_input_file.url), but since during initial submission of the file this would always be NONE. So should I change it to self.submission.submission_input_file or keep it as None?

File: apps/jobs/views.py (hunk 5 above)
Comment (savish28) [reply apps/jobs/views.py]: @KhalidRmb Done the required change.

File: tests/unit/jobs/test_views.py (hunk 7 above)
Comment (KhalidRmb) [reply tests/unit/jobs/test_views.py:672]: > @KhalidRmb, Ideally since it is a file it should be "http://testserver%s"% (self.submission.submission_input_file.url), but since during initial submission of the file this would always be NONE. So should I change it to self.submission.submission_input_file or keep it as None? Same is the case with stderr/stoud file they are also checked as None.

Makes sense. None is fine for now

File: apps/challenges/models.py [hunk 8]
```diff
@@ -102,6 +102,11 @@ def __init__(self, *args, **kwargs):
     is_docker_based = models.BooleanField(
         default=False, verbose_name="Is Docker Based", db_index=True
     )
+    is_static_dataset_docker_based_challenge = models.BooleanField(
+        default=False,
+        verbose_name="Is Static Code Docker Based",
```
Comment (Ram81) [inline apps/challenges/models.py:107]: Change to `Is static dataset docker based`

File: apps/jobs/views.py [hunk 9]
```diff
@@ -483,6 +486,18 @@ def change_submission_data_and_visibility(
     if serializer.is_valid():
         serializer.save()
         response_data = serializer.data
+        if (
+            request.FILES.get("submission_input_file")
+            and challenge.is_static_dataset_docker_based_challenge
```
Comment (Ram81) [inline apps/jobs/views.py:491]: @savish28 did you test this API by making a call with `submission_input_file`?

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: apps/jobs/views.py (hunk 9 above)
Comment (savish28) [reply apps/jobs/views.py:491]: @Ram81 Yeah I have checked both the file submission and also the message by making a static code upload challenge and a normal challenge by Postman.

File: apps/challenges/models.py (hunk 8 above)
Comment (savish28) [reply apps/challenges/models.py]: Done.

**Comment by Ram81** [comment]
@vkartik97 @KhalidRmb can you verify the changes locally once.

File: apps/challenges/models.py [hunk 10]
```diff
@@ -102,6 +102,11 @@ def __init__(self, *args, **kwargs):
     is_docker_based = models.BooleanField(
         default=False, verbose_name="Is Docker Based", db_index=True
     )
+    is_static_dataset_docker_based_challenge = models.BooleanField(
+        default=False,
+        verbose_name="Is static dataset docker based",
```
Comment (Ram81) [inline apps/challenges/models.py:107]: @savish28 please change it to `Is Static Dataset Docker Based`

**Review by Ram81 — CHANGES_REQUESTED** [review]
@savish28 minor change. Rest of it LGTM. 

@RishabhJain2018 can you take a look

**Review by Ram81 — APPROVED** [review]

File: apps/challenges/admin.py [hunk 11]
```diff
@@ -52,6 +52,7 @@ class ChallengeAdmin(ImportExportTimeStampedAdmin):
         "featured",
         "created_at",
         "is_docker_based",
+        "is_static_dataset_docker_based_challenge",
```
Comment (RishabhJain2018) [inline apps/challenges/admin.py:55]: `is_static_dataset_docker_based_challenge` -> `is_static_dataset_docker_based`

File: apps/jobs/sender.py [hunk 12]
```diff
@@ -65,6 +65,7 @@ def publish_submission_message(message):
             - "phase_pk": int
             - "submission_pk": int
             - "submitted_image_uri": str, (only available when the challenge is a code upload challenge)
+            - "is_static_code_upload_submission": bool
```
Comment (RishabhJain2018) [inline apps/jobs/sender.py:68]: `is_static_code_upload_submission` -> `is_static_dataset_docker_based`

File: apps/jobs/utils.py [hunk 13]
```diff
@@ -211,6 +211,7 @@ def handle_submission_rerun(submission, updated_status):
         "challenge_pk": submission.challenge_phase.challenge.pk,
         "phase_pk": submission.challenge_phase.pk,
         "submission_pk": submission.pk,
+        "is_static_code_upload_submission": False,
```
Comment (RishabhJain2018) [inline apps/jobs/utils.py:214]: `is_static_dataset_docker_based`

File: apps/jobs/utils.py [hunk 14]
```diff
@@ -224,6 +225,11 @@ def handle_submission_rerun(submission, updated_status):
             message["submitted_image_uri"] = response.json()[
                 "submitted_image_uri"
             ]
+            if (
+                submission.challenge_phase.challenge.is_static_dataset_docker_based_challenge
+            ):
+                message["is_static_code_upload_submission"] = True
```
Comment (RishabhJain2018) [inline apps/jobs/utils.py:231]: `is_static_dataset_docker_based `

File: apps/jobs/views.py [hunk 15]
```diff
@@ -351,13 +351,16 @@ def challenge_submission(request, challenge_id, challenge_phase_id):
         message = {
             "challenge_pk": challenge_id,
             "phase_pk": challenge_phase_id,
+            "is_static_code_upload_submission": False,
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:354]: `is_static_dataset_docker_based`

File: apps/jobs/views.py [hunk 16]
```diff
@@ -351,13 +351,16 @@ def challenge_submission(request, challenge_id, challenge_phase_id):
...
         }
         if challenge.is_docker_based:
             try:
                 file_content = json.loads(request.FILES["input_file"].read())
                 message["submitted_image_uri"] = file_content[
                     "submitted_image_uri"
                 ]
+                if challenge.is_static_dataset_docker_based_challenge:
+                    message["is_static_code_upload_submission"] = True
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:363]: `is_static_dataset_docker_based`

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Review by RishabhJain2018 — COMMENTED** [review]
Please make change to the variable name at all places in the PR.

**Comment by savish28** [comment]
> Please make change to the variable name at all places in the PR.

Thanks for the suggestion. Done :+1:

**Review by RishabhJain2018 — APPROVED** [review]
